crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -139,6 +159,52 @@ where
             is_valid
         })
     }
+
+    /// Watch block reorgs and update filters accordingly
+    async fn watch_reorgs(&self, mut notifications: CanonStateNotificationStream) {
+        while let Some(notification) = notifications.next().await {
+            if let CanonStateNotification::Reorg { old, .. } = notification {
+                self.update_reorg(old.blocks()).await;
+            }
+        }
+    }
+
+    /// update a reorg block for all active filters
+    async fn update_reorg(&self, old_blocks: &BTreeMap<BlockNumber, SealedBlockWithSenders>) {
+        let reorg_blocks: HashMap<BlockHash, BlockNumber> =
+            old_blocks.iter().map(|(k, v)| (v.header.hash(), *k)).collect();
+
+        for active_filter in self.active_filters().inner.lock().await.values_mut() {
+            if let FilterKind::Log(filter) = &active_filter.kind {
+                match filter.block_option {
+                    FilterBlockOption::AtBlockHash(block_hash) => {
+                        if let Some(block_number) = reorg_blocks.get(&block_hash) {
+                            active_filter.reorg_blocks.write().insert(block_hash, *block_number);
```

Comment:
this update is slightly flawed because this doesn't actually preserve the reorged logs.

the solution here should a be a lot simpler, because all this has to do is filter those logs from the reorged chain that
a) match the filter
b) have been previously emitted (this info we get from looking at ActibeFilter's block value)
https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/eth/filter.rs#L671-L672

so this should actually be just a Vec of `Log`
https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/eth/filter.rs#L677-L678
which represent "logs that we previously emitted but are now reorged (removed = true)"